msdm/core/utils/hashdictionary.py

```diff
@@ -0,0 +1,104 @@
+import inspect
+
+class HashDictionary(dict):
+    """Dictionary with a function for hashing keys"""
+    def __init__(self, hash_function=None, *args, **kwargs):
+        if hash_function is None:
+            hash_function = lambda x: x
+        self.hash_function = hash_function
+        self._encoded_keys = {} #note: this isn't doing garbage collection
```

Comment:
Stray thought; if you permit optional specification of a dehashing function (mathematically not possible of course for a proper hash, but probably possible for our functions which return hashable representations), then you can optionally avoid storing this mapping. Potential performance/memory savings. Especially nice for the case of already-hashable states, since you avoid storing this & can "dehash" through the identity function.

Reply:
set this up!